api/views.py

```diff
@@ -0,0 +1,161 @@
+import datetime
+from datetime import timedelta, datetime
+
+from django.contrib.auth.models import User, Group
+from django.db.models import Sum
+from django.shortcuts import get_object_or_404
+from rest_framework import viewsets
+from rest_framework.decorators import action
+from rest_framework.response import Response
+
+from api.serializers import UserSerializer, GroupSerializer
+from .models import Employee, Income, Outcome, Event
+from .serializers import EmployeeSerializer, EventMiniSerializer, IncomeSerializer, OutcomeSerializer, EventSerializer
+
+
+class UserViewSet(viewsets.ModelViewSet):
+    queryset = User.objects.all().order_by('-id')
+    serializer_class = UserSerializer
+
+
+class GroupViewSet(viewsets.ModelViewSet):
+    queryset = Group.objects.all()
+    serializer_class = GroupSerializer
+
+
+class EmployeeViewSet(viewsets.ModelViewSet):
+    queryset = Employee.objects.all()
+    serializer_class = EmployeeSerializer
+
+    def list(self, request, *args, **kwargs):
+        number_of_employees = request.GET.get('number_of_employees')
+        employees = Employee.objects.order_by('-salary')
+
+        if number_of_employees:
+            employees = employees[:int(number_of_employees)]
+
+        serializer = self.get_serializer(employees, many=True)
+        return Response(serializer.data)
+
+    @action(detail=True, methods=['get'])
+    def fire(self, request, *args, **kwargs):
+        employee = self.get_object()
+        employee.hired = False
+        employee.save(update_fields=['hired'])
+        serializer = EmployeeSerializer(employee, many=False)
+        serializer.is_valid(raise_exception=True)
+        self.perform_update(serializer)
+        return Response(serializer.data)
+
+    @action(detail=True, methods=['get'])
+    def hire_again(self, request, *args, **kwargs):
+        employee = self.get_object()
+        employee.hired = True
+        employee.save(update_fields=['hired'])
+        serializer = EmployeeSerializer(employee, many=False)
+        serializer.is_valid(raise_exception=True)
```

Comment:
We're validating whether the data is valid... after we have already saved it to database. Moreover, `self.perform_update(serializer)` will save it once again to the database.

```python
employee = self.get_object()
employee.hired = True
serializer = EmployeeSerializer(employee)
serializer.is_valid(raise_exception=True)
self.perform_update(serializer)
return Response(serializer.data)
```

This will do the work as well. We'll set `employee.hired` but only in the memory, not in the database. Then we will serializer this changed object and only in case that the serializer does not raise and error, then we will `perform_update` and this will save it to database as well.

Reply:
Also, fix this in all places, not just here.

Reply:
done